README.md
# ubuntu

1. [Overview](#overview)
1. [Description](#description)
1. [Tags](#tags)
1. [Setup](#setup)
1. [Usage](#usage)
1. [Limitations](#limitations)
1. [Development](#development)
1. [Miscellaneous](#miscellaneous)

## Overview

Ubuntu is  a Debian-based  free operating  system (OS)  for your  computer.  An
operating system  is the  set of  basic programs and  utilities that  make your
computer run.

[ubuntu.com](https://www.ubuntu.com/)

## Description

Use this script to build your own base system.

We've included the last ca-certificates files  in the repository to ensure that
all of our images are accurates.

## Tags

Supported tags.

- 12.04, precise
- 14.04, trusty
- 16.04, xenial
- 18.04, bionic
- 20.04, focal, latest

## Setup

On Debian you need sudo permissions and the following packages:

```bash
# if you build on wheezy please use backports version of debootstrap
sudo apt-get install debootstrap ubuntu-archive-keyring
```

On Devuan you need sudo permissions and the following packages:

```bash
sudo apt-get -qq -y install debootstrap ubuntu-archive-keyring
```

On Ubuntu you need sudo permissions and the following packages:

```bash
sudo apt-get install debootstrap
```

You also need to be in the docker group to use Docker.

```bash
sudo usermod -a -G docker ${USER}
```

Finally you need to login on Docker Hub.

```bash
docker login
```

## Usage

You first need to choose which dist between precise (12.04), trusty (14.04),
xenial (16.04) and bionic (18.08) you want  (bionic will  be the  'latest' tag)
and you  need to choose you user (or organization) name on Docker Hub.

Show help.

```bash
./build.sh -h
```

Build your own Ubuntu image (eg. xenial).

```bash
./build.sh -d xenial -u vptech
```

Build your own Ubuntu image (eg. bionic) and push it on the Docker Hub.

```bash
./build.sh -d bionic -u vptech -p
```

## Limitations

Only work on Debian, Devuan and Ubuntu.

## Development

Please read carefully [CONTRIBUTING.md](CONTRIBUTING.md) before making a merge
request.

## Miscellaneous

```
    ╚⊙ ⊙╝
  ╚═(███)═╝
 ╚═(███)═╝
╚═(███)═╝
 ╚═(███)═╝
  ╚═(███)═╝
   ╚═(███)═╝
```


CHANGELOG
## [1.0.1](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/compare/1.0.0...1.0.1) (2021-12-13)


### Bug Fixes

* semantic changelog ([74effd3](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/74effd3aeb958cc3abd9946b74383299c17ed4cd))


### Chores

* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/markdown-lint to v1.2.3 ([f318703](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/f318703159faae4875717e05216f104d407ed98e))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/markdown-lint to v1.3.0 ([eb4e8b5](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/eb4e8b545036dcf6f9f9b56e37762ed8fa9d4934))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/markdown-lint to v1.4.0 ([83c6b8f](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/83c6b8fae374d29dffbddfe3e2882e4a43253714))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/semantic-release to v3.10.0 ([94c50f9](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/94c50f9e2e3903da0af6a75a1326fb30f592a355))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/semantic-release to v3.3.0 ([034b35b](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/034b35bb51e7bed60472d70a194daa8f3c3dd780))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/semantic-release to v3.5.0 ([fb0712d](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/fb0712d6cb156e15afa049ca266e69a55f840790))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/semantic-release to v3.6.0 ([541fc55](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/541fc5568c089ddb977f9b349a7d6c781e2d94d9))
* **deps:** update dependency veepee/vptech/sre/iac/ci-cd/semantic-release to v3.9.0 ([1b283a9](https://git.vptech.eu/veepee/vptech/sre/iac/docker-images/ubuntu/commit/1b283a999965d7d8955d1f5c2754148b85836aa9))


build.sh
#!/usr/bin/env bash
# shellcheck disable=SC1039
# shellcheck disable=SC2034

set -e

PATH='/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'

arch='amd64'
oldstable='bionis'
stable='focal'
testing='focal'
version='4.1'

function usage()
{
    cat <<EOF

NAME:
   build.sh - Docker images' builder of Ubuntu.

USAGE:
   build.sh -d <dist>

OPTIONS:
   -h, --help           Show help

   -d, --dist           Choose Ubuntu distribution
                        eg: precise, trusty, xenial, bionic, focal

   -e, --extra-packages space separated list of extra packages
                        eg: -e foo bar baz
                        default: none

   -T  --extra-tag      Add a tag suffix to the created one
                        eg: buster-stable => buster-stable-{tag}
                        default: none

   -t, --timezone       Choose your preferred timezone
                        default: Europe/Amsterdam

   -u, --user           Docker Hub username or organisation
                        default: none

   -p, --push           Docker Hub push
                        default: no

   -l, --latest         Force the "latest"
                        default: bionic

   -v, --verbose        Verbose mode

   -V, --version        Show version

VERSION:
   docker-ubuntu version: ${version}

EOF
}

function docker_bootstrap()
{
    # variables
    image="/tmp/image-${distname}-${arch}"
    include="${include},\
             apt-utils,\
             ca-certificates,\
             curl,\
             git,\
             locales"
    exclude="debconf-i18n,\
             dmsetup,\
             git-man,\
             info,\
             initramfs-tools,\
             man-db,\
             manpages"

    echo '-- bootstrap' 1>&3

    if [ "$(id -u)" -ne 0 ]
    then
        sudo='sudo'
    fi

    # clean old image
    if [ -d "/tmp/image-${distname}-${arch}" ]
    then
        ${sudo} rm -fr "${image}"
    fi

    # create minimal debootstrap image
    if [ ! -f "/usr/share/debootstrap/scripts/${distname}" ] || \
       [ ! -h "/usr/share/debootstrap/scripts/${distname}" ]
    then
        echo "File /usr/share/debootstrap/scripts/${distname} is missing." 1>&3
        echo "1.) did you install backports version of debootstrap ?" 1>&3
        echo "2.) ln -s gusty /usr/share/debootstrap/scripts/${distname}" 1>&3
        exit 1
    else
        echo " * debootstrap ${image}" 1>&3
        ${sudo} debootstrap \
                --arch="${arch}" \
                --include="${include}" \
                --exclude="${exclude}" \
                --variant=minbase \
                "${distname}" \
                "${image}" \
                "${mirror}"
        if [ ${?} -ne 0 ]
        then
            echo "There is an issue Issue with debootstrap." 1>&3
            echo "Please run again with -v." 1>&3
            exit 1
        fi
    fi

    # create /etc/default/locale
    echo ' * /etc/default/locale' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/default/locale"
LANG=en_US.UTF-8
LANGUAGE=en_US.UTF-8
LC_COLLATE=en_US.UTF-8
LC_ALL=en_US.UTF-8
EOF

    # create /etc/timezone
    echo ' * /etc/timezone' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/timezone"
${timezone}
EOF

    # create /etc/resolv.conf
    echo ' * /etc/resolv.conf' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/resolv.conf"
nameserver 8.8.4.4
nameserver 8.8.8.8
EOF

    # create /etc/apt/sources.list
    echo ' * /etc/apt/sources.list' 1>&3
    ${sudo} cp -r "sources.list/${distname}" "${image}/etc/apt/sources.list"

    if [ "${distname}" != 'lucid' ]
    then
       # create /etc/dpkg/dpkg.cfg.d/disable-doc
       echo ' * /etc/dpkg/dpkg.cfg.d/disable-doc' 1>&3
       cat <<EOF | \
           ${sudo} tee "${image}/etc/dpkg/dpkg.cfg.d/disable-doc"
path-exclude /usr/share/doc/*
path-include /usr/share/doc/*/copyright
path-exclude /usr/share/man/*
path-exclude /usr/share/groff/*
path-exclude /usr/share/info/*
path-exclude /usr/share/lintian/*
path-exclude /usr/share/linda/*
EOF
    fi

    # create /etc/apt/apt.conf.d/force-ipv4
    echo ' * /etc/apt/apt.conf.d/force-ipv4' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/apt/apt.conf.d/force-ipv4"
Acquire::ForceIPv4 "true";
EOF

    # create /etc/apt/apt.conf.d/disable-auto-install
    echo ' * /etc/apt/apt.conf.d/disable-auto-install' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/apt/apt.conf.d/disable-auto-install"
APT::Install-Recommends "0";
APT::Install-Suggests "0";
EOF

    # create /etc/apt/apt.conf.d/disable-cache
    echo ' * /etc/apt/apt.conf.d/disable-cache' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/apt/apt.conf.d/disable-cache"
Dir::Cache::pkgcache "";
Dir::Cache::srcpkgcache "";
EOF

    # create /etc/apt/apt.conf.d/force-conf
    echo ' * /etc/apt/apt.conf.d/force-conf' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/apt/apt.conf.d/force-conf"
Dpkg::Options {
   "--force-confnew";
   "--force-confmiss";
}
EOF

    # create /etc/apt/apt.conf.d/disable-languages
    echo ' * /etc/apt/apt.conf.d/disable-languages' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/etc/apt/apt.conf.d/disable-languages"
Acquire::Languages "none";
EOF

    # create /usr/bin/apt-clean
    echo ' * /usr/bin/apt-clean' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/usr/bin/apt-clean"
#!/bin/bash

# Please read https://wiki.ubuntu.com/ReducingDiskFootprint

find /usr/share/doc     -type f ! -name copyright -delete
find /usr/share/i18n    -type f -delete
find /usr/share/locale  -type f -delete
find /usr/share/man     -type f -delete
find /var/cache/apt     -type f -delete
find /var/lib/apt/lists -type f -delete

apt-get autoclean       -qq -y
apt-get autoremove      -qq -y
apt-get clean           -qq -y
# EOF
EOF
    ${sudo} chmod 755 "${image}/usr/bin/apt-clean"

    # create /usr/sbin/policy-rc.d
    echo ' * /usr/sbin/policy-rc.d' 1>&3
    cat <<EOF | \
        ${sudo} tee "${image}/usr/sbin/policy-rc.d"
#!/bin/bash

exit 101
# EOF
EOF
    ${sudo} chmod 755 "${image}/usr/sbin/policy-rc.d"

    if [ "${distname}" != 'lucid' ]
    then
        # update initctl
        echo ' * /sbin/initctl' 1>&3
        ${sudo} chroot "${image}" bash -c \
                "dpkg-divert --local --rename --add /sbin/initctl && \
                 ln -sf /bin/true /sbin/initctl" 2>&1
    fi

    # mount
    ${sudo} mount --bind /dev     "${image}/dev"
    ${sudo} mount --bind /dev/pts "${image}/dev/pts"
    ${sudo} mount --bind /proc    "${image}/proc"
    ${sudo} mount --bind /sys     "${image}/sys"

    # update root certificates
    ${sudo} mkdir -p "${image}/usr/local/share/"
    ${sudo} cp -r ca-certificates "${image}/usr/local/share/"

    # upgrade (without output...)
    echo ' * apt-get upgrade' 1>&3
    ${sudo} chroot "${image}" bash -c \
            "export DEBIAN_FRONTEND=noninteractive && \
             update-ca-certificates -f && \
             apt-get update && \
             apt-get upgrade -y && \
             apt-get dist-upgrade -y && \
             apt-clean"

    # unmount
    ${sudo} umount "${image}/dev/pts"
    ${sudo} umount "${image}/dev"
    # ${sudo} umount "${image}/proc"
    ${sudo} umount "${image}/sys"

    # create archive
    if [ -f "${image}.tar" ]
    then
        ${sudo} rm "${image}.tar"
    fi
    ${sudo} tar -C "${image}" -c -f "${image}.tar" --numeric-owner .
}

# create images from bootstrap archive
function docker_import()
{
    echo "-- docker import from ${image}" 1>&3
    mount -t proc proc /proc
    docker import "${image}.tar" "${user}ubuntu:${distname}${tag}"
    docker run "${user}ubuntu:${distname}${tag}" \
           echo " * build ${user}ubuntu:${distname}${tag}" 1>&3
    docker tag "${user}ubuntu:${distname}${tag}" "${user}ubuntu:${distid}${tag}"
    docker run "${user}ubuntu:${distid}${tag}" \
           echo " * build ${user}ubuntu:${distid}${tag}" 1>&3

    for import in latest oldstable stable testing
    do
        if [ "${distname}" = "${!import}" ]
        then
            docker tag "${user}ubuntu:${distname}${tag}" "${user}ubuntu:${import}${tag}"
            docker run "${user}ubuntu:${import}${tag}" \
                   echo " * build ${user}ubuntu:${import}${tag}" 1>&3
        fi
    done
}

# push image to docker hub
function docker_push()
{
    echo "-- docker push" 1>&3
    echo " * push ${user}ubuntu:${distname}${tag}" 1>&3
    docker push "${user}ubuntu:${distname}${tag}"
    echo " * push ${user}ubuntu:${distid}${tag}" 1>&3
    docker push "${user}ubuntu:${distid}${tag}"

    for push in latest oldstable stable testing
    do
        if [ "${distname}" = "${!push}"  ]
        then
            echo " * push ${user}ubuntu:${push}${tag}" 1>&3
            docker push "${user}ubuntu:${push}${tag}"
        fi
    done
}

while getopts 'hd:e:T:t:u:plvV' OPTIONS
do
    case ${OPTIONS} in
        h)
            # -h / --help
            usage
            exit 0
            ;;
        d)
            # -d / --dist
            dist=${OPTARG}
            ;;
        e)
            # -e / --extra-packages
            extra=${OPTARG}
            ;;
        T)
            # -T / --tag
            tag=${OPTARG}
            ;;
        t)
            # -t / --timezone
            timezone=${OPTARG}
            ;;
        u)
            # -u / --user
            user="${OPTARG}/"
            ;;
        p)
            # -p / --push
            push='true'
            ;;
        l)
            # -l / --latest
            latest=${OPTARG}
            ;;
        v)
            # -v / --verbose
            verbose='true'
            ;;
        V)
            # -V / --version
            echo "${version}"
            exit 0
            ;;
        *)
            usage
            exit 1
            ;;
    esac
done

if [ ! -x "$(command -v sudo)" ]
then
    echo "Please install sudo (see README.md)"
    exit 1
fi

if [ ! -x "$(command -v debootstrap)" ]
then
    echo "Please install debootstrap (see README.md)"
    exit 1
fi

# -e / --extra-packages
if [ -z "${extra}" ]
then
    extra=''
fi

# -d / --dist
if [ -n "${dist}" ]
then
    case ${dist} in
        lucid|10.04|10.04-lts)
            distname='lucid'
            distid='10.04'
            mirror='http://old-releases.ubuntu.com/ubuntu'
            include="apt-transport-https,git-core,gpgv,${extra}"
            ;;
        precise|12.04|12.04-lts)
            distname='precise'
            distid='12.04'
            mirror='http://mirror.vpgrp.io/ubuntu'
	        include="apt-transport-https,git-core,${extra}"
            ;;
        trusty|14.04|14.04-lts)
            distname='trusty'
            distid='14.04'
            mirror='http://mirror.vpgrp.io/ubuntu'
	        include="apt-transport-https,git-core,${extra}"
            ;;
        xenial|16.04|16.04-lts)
            distname='xenial'
            distid='16.04'
            mirror='http://mirror.vpgrp.io/ubuntu'
	        include="apt-transport-https,git-core,${extra}"
            ;;
        bionic|18.04|18.04-lts)
            distname='bionic'
            distid='18.04'
            mirror='http://mirror.vpgrp.io/ubuntu'
            include="${extra}"
            ;;
        focal|20.04|20.04-lts)
            distname='focal'
            distid='20.04'
            mirror='http://mirror.vpgrp.io/ubuntu'
            include="${extra}"
            ;;
        *)
            usage
            exit 1
            ;;
    esac
else
    usage
    exit 1
fi

# -t / --timezone
if [ -z "${timezone}" ]
then
    timezone='Europe/Amsterdam'
fi


# -T / --tag
if [ -z "${tag}" ]
then
    tag=''
else
    tag="-${tag}"
fi

# -u / --user
if [ -z "${user}" ]
then
    user=''
fi

# -l / --latest
if [ -z "${latest}" ]
then
    latest='focal'
fi

# -v / --verbose
if [ -z "${verbose}" ]
then
    exec 3>&1
    exec 1>/dev/null
    exec 2>/dev/null
else
    exec 3>&1
fi

docker_bootstrap
docker_import

if [ -n "${push}" ]
then
    docker_push
fi
# EOF
